src/Microsoft.Net.Http.Headers/NameValueHeaderValue.cs

```diff
@@ -142,6 +143,32 @@ public override bool Equals(object obj)
             }
         }
 
+        public StringSegment GetDecodedValue()
+        {
+            if (!HeaderUtilities.IsQuoted(_value))
+            {
+                return _value;
+            }
+            return HeaderUtilities.DecodeQuotedString(_value);
+        }
+
+        public void SetAndEncodeValue(StringSegment value)
+        {
+            if (StringSegment.IsNullOrEmpty(value) || (GetValueLength(value, 0) == value.Length))
+            {
+                _value = value;
+            }
+            else
+            {
+                var encodedValue = HeaderUtilities.EscapeAsQuotedString(value);
+                if (GetValueLength(encodedValue, 0) != encodedValue.Length)
```

Comment:
This isn't strong enough because `GetValueLength()` does not confirm quoted strings contain only valid characters. Further, the parsing routines explicitly ignore invalid quoted pairs.